staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,266 @@
+	// This should be executed after restarting the watch (if applicable) to ensure no file event will be missing.
+	defer d.queue.Add(workqueueKey)
+
+	// return if file has not removed or renamed.
+	if event.Op&(fsnotify.Remove|fsnotify.Rename) == 0 {
+		return nil
+	}
+
+	if err := watcher.Remove(d.filePath); err != nil {
+		klog.V(2).InfoS("Failed to remove file watch, it may have been deleted", "file", d.filePath, "err", err)
+	}
+	if err := watcher.Add(d.filePath); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", d.filePath, err)
+	}
+
+	return nil
+}
+
+// runWorker to process file content
+func (d *DynamicKMSEncryptionConfigContent) runWorker() {
+	for d.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem processes file content when there is a message in the queue.
+func (d *DynamicKMSEncryptionConfigContent) processNextWorkItem() bool {
+	// key here is dummy item in the queue to trigger file content processing.
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+
+	err := d.loadEncryptionConfig()
+	if err == nil {
+		d.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// add dummy item back to the queue to trigger file content processing.
+	d.queue.AddRateLimited(key)
+
+	return true
+}
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (d *DynamicKMSEncryptionConfigContent) loadEncryptionConfig() error {
+	encryptionConfig, err := d.readEncryptionFileContent(d.filePath)
+	if err != nil {
+		return err
+	}
+	if len(encryptionConfig) == 0 {
+		return fmt.Errorf("encryption config file is empty %q", d.filePath)
+	}
+
+	// compute hash for the encryption config file content.
+	encryptionConfigHash := d.computeHash(encryptionConfig)
+
+	// check if encryptionConfig is different from the current. Do nothing if they are the same.
+	if !d.hasEncryptionConfigChanged(encryptionConfigHash) {
+		return nil
+	}
+
+	err = d.processEncryptionConfig()
+	if err != nil {
+		return err
+	}
+
+	// update local copy of recent config content once processing is successful.
+	d.encryptionConfig.Store(encryptionConfigHash)
+	klog.V(2).InfoS("Loaded new kms encryption config content", "name", d.name)
+
+	return nil
+}
+
+func (d *DynamicKMSEncryptionConfigContent) readEncryptionFileContent(filePath string) ([]byte, error) {
+	return os.ReadFile(d.filePath)
+}
+
+// processEncryptionConfig processes the encryption config.
+func (d *DynamicKMSEncryptionConfigContent) processEncryptionConfig() error {
+	// get context to close the old transformers.
+	ctx, closeTransformers := wait.ContextForChannel(d.stopCh)
+
+	transformerOverrides, _, err := encryptionconfig.LoadEncryptionConfig(d.filePath, ctx.Done())
+	if err != nil {
+		klog.Warning("KMS encryption configuration is not valid.", err)
+		// close transformers to avoid leaking resources.
+		closeTransformers()
+		return err
+	}
+
```

Comment:
should we also health check the new transformers before making them effective?

Reply:
added health check before calling `Set` to make sure new transformers are healthy.